### PR TITLE
[SYCLomatic][part1] Implement cmake script migration framework

### PR DESCRIPTION
**Pull request by tomflinda**
1. Add two options:
  --migrate-cmake-script         - migrate the cmakfile build script. Default: off.
  --migrate-cmake-script-only - Only migrate the cmakfile build script. Default: off.
  
2. Implement cmake migration framework to do cmake syntax migration

### CODE REVIEW

File: clang/lib/DPCT/DPCT.cpp [hunk 1]
```diff
@@ -542,6 +543,156 @@ void parseFormatStyle() {
...
+    }
+    // Skip hidden folder or file whose name begins with ".".
+    if (IsHidden) {
+      continue;
+    }
+
+    if (Iter->type() == fs::file_type::directory_file) {
+      const auto Path = Iter->path();
+      SmallString<512> OutDirectory = llvm::StringRef(Path);
+      if (fs::exists(OutDirectory + "/CMakeFiles") &&
```
Comment (zhiweij1) [inline clang/lib/DPCT/DPCT.cpp]: I suggest to use path::append() instead of using "+"

File: clang/lib/DPCT/PatternRewriter.cpp [hunk 2]
```diff
@@ -474,13 +569,63 @@ bool fixLineEndings(const std::string &Input, std::string &Output) {
...
+      std::transform(Str.begin(), Str.end(), Str.begin(),
+                     [](unsigned char c) { return std::tolower(c); });
+      if (cmake_commands.find(Str) != cmake_commands.end()) {
+        for (int Idx = Begin; Idx < End; Idx++) {
+          Line[Idx] = Str[Idx - Begin];
+        }
+      }
+    }
+
+    OutputStream << Line << "\n";
```
Comment (zhiweij1) [inline clang/lib/DPCT/PatternRewriter.cpp:600]: getNL() ?

File: clang/lib/DPCT/PatternRewriter.cpp [hunk 3]
```diff
@@ -474,13 +569,63 @@ bool fixLineEndings(const std::string &Input, std::string &Output) {
...
+  }
+
+  return OutputStream.str();
+}
+
+int skipCmakeComments(std::ostream &OutputStream, const std::string &Input,
+                      int Index) {
+  const int Size = Input.size();
+  if (Input[Index] == '#') {
+    for (; Index < Size && Input[Index] != '\n'; Index++) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/PatternRewriter.cpp:610]: what will happen to Windows style \r\n ?

File: clang/lib/DPCT/PatternRewriter.cpp (hunk 2 above)
Comment (tomflinda) [reply clang/lib/DPCT/PatternRewriter.cpp:600]: Cannot use genNL,  here "\n" is hard-coded separator.

File: clang/lib/DPCT/DPCT.cpp (hunk 1 above)
Comment (tomflinda) [reply clang/lib/DPCT/DPCT.cpp]: okay

File: clang/lib/DPCT/PatternRewriter.cpp (hunk 3 above)
Comment (tomflinda) [reply clang/lib/DPCT/PatternRewriter.cpp:610]: Input has been processed to conver "\r\n" to "\n" before this function call, so only need to check "\n"

**Review by zhiweij1 — APPROVED** [review]

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 4]
```diff
@@ -388,6 +388,13 @@ DPCT_NON_ENUM_OPTION(DPCT_OPT_TYPE(llvm::cl::opt<std::string>), BuildScriptFile,
                     "Default name: Makefile.dpct."),
         llvm::cl::value_desc("file"), llvm::cl::cat(DPCTCat), llvm::cl::Optional)
 
+DPCT_NON_ENUM_OPTION(DPCT_OPT_TYPE(llvm::cl::opt<bool>), MigrateCmakeScript, "migrate-cmake-script",
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc:391]: Please add experimental tag util the feature is ready for end user.

File: clang/lib/DPCT/DPCT.cpp [hunk 5]
```diff
@@ -542,6 +543,156 @@ void parseFormatStyle() {
   DpctGlobalInfo::setCodeFormatStyle(Style);
 }
 
+static std::string getCmakeBuildPathFromInRoot(StringRef InRoot,
```
Comment (zhimingwang36) [inline clang/lib/DPCT/DPCT.cpp]: could you put all cmake script migration logic into a new cpp/hpp files?

File: clang/lib/DPCT/DPCT.cpp [hunk 6]
```diff
@@ -655,7 +806,8 @@ int runDPCT(int argc, const char **argv) {
     dpctExit(MigrationErrorPathTooLong);
   }
   if (AnalysisScope.size() >= MAX_PATH_LEN - 1) {
-    DpctLog() << "Error: --analysis-scope-path '" << AnalysisScope << "' is too long\n";
+    DpctLog() << "Error: --analysis-scope-path '" << AnalysisScope
```
Comment (zhimingwang36) [inline clang/lib/DPCT/DPCT.cpp]: please keep foot print as small as possible.

File: clang/lib/DPCT/DPCT.cpp [hunk 7]
```diff
@@ -1090,6 +1273,33 @@ int runDPCT(int argc, const char **argv) {
     parseFormatStyle();
   }
 
+  if (MigrateCmakeScriptOnly) {
+    auto CmakeScriptLists = OptParser->getSourcePathList();
```
Comment (zhimingwang36) [inline clang/lib/DPCT/DPCT.cpp]: Please consider put the functions, to simplify the DPCT.cpp main framwork.

File: clang/lib/DPCT/Error.cpp [hunk 8]
```diff
@@ -183,6 +183,21 @@ void ShowStatus(int Status, std::string Message) {
   case InterceptBuildError:
     StatusString = "Error: Call to intercept-build failed";
     break;
+  case MigrationErrorCMakeScriptPathInvalid:
+    StatusString = "Error: Cmake Script Path invalid.\n";
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Error.cpp]: Path of Cmake Script is invalid.

File: clang/lib/DPCT/PatternRewriter.cpp [hunk 9]
```diff
@@ -37,6 +37,97 @@ struct MatchResult {
   std::unordered_map<std::string, std::string> Bindings;
 };
 
+extern llvm::cl::opt<bool> MigrateCmakeScriptOnly;
```
Comment (zhimingwang36) [inline clang/lib/DPCT/PatternRewriter.cpp:40]: Please  put the logic into separate file which is specific to cmake build scripts migration.

File: clang/include/clang/DPCT/DPCTOptions.inc (hunk 4 above)
Comment (tomflinda) [reply clang/include/clang/DPCT/DPCTOptions.inc:391]: Okay.

File: clang/lib/DPCT/DPCT.cpp (hunk 5 above)
Comment (tomflinda) [reply clang/lib/DPCT/DPCT.cpp]: Okay

File: clang/lib/DPCT/DPCT.cpp (hunk 6 above)
Comment (tomflinda) [reply clang/lib/DPCT/DPCT.cpp]: Okay

File: clang/lib/DPCT/DPCT.cpp (hunk 7 above)
Comment (tomflinda) [reply clang/lib/DPCT/DPCT.cpp]: Okay

File: clang/lib/DPCT/Error.cpp (hunk 8 above)
Comment (tomflinda) [reply clang/lib/DPCT/Error.cpp]: ok

File: clang/lib/DPCT/PatternRewriter.cpp (hunk 9 above)
Comment (tomflinda) [reply clang/lib/DPCT/PatternRewriter.cpp:40]: The option imported here is used to differentiate the logic specific for cmake script migration in pattern rewriter,  it is better to keep it here temporarily. I will change it later in the next part2 pr.

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**